[Sebastian Lee] iP

sebbycake opened the pull request:
# Sync-Me Sebby

> “Focus on being productive instead of busy.” – Tim Ferriss ([source](https://www.goodreads.com/quotes/834320-focus-on-being-productive-instead-of-busy))

# Description

In this pull request, I have added several features to the base Duke application. These features include:

- Ability to add, remove and find action items divided into 3 categories - **todos**, **deadline** and **event**
- Ability to mark action item as complete
- Ability to load and write list of action items to hard disk in the form of a `.txt` file
- _Extensive_ implementation of OOP principles
- Handling errors, especially of user inputs
- Added unit tests


## Types of change

- [x] New features
- [x] Refactoring with OOP principles


# Testing

To run JUnit tests,
1. Open the project using an IDE like IntelliJ. 
2. Go to the `test` directory.
3. Right click, and click `Run all tests`.


To run automated text testing, open your terminal:
```bash
cd <project-root-directory>/text-ui-test

# for Linux/macOS
./runtest.sh

# for Windows
runtest.bat
```

- [x] JUnit Tests
- [x] Automated text testing using bash script


# Checklist:

- [ ] My code follows the Java style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

Do try out the app :smile: !

Commit:
Add Gradle support

sebbycake committed:
Added greeting, echoing and exiting functionalities of bot

sebbycake committed:
Added the ability to save and display tasks inputted

sebbycake committed:
Added feature to mark tasks as completed

sebbycake committed:
Added several subtypes of Task with different status messages

sebbycake committed:
Added test case to do automated testing

sebbycake committed:
Handled errors from user input

sebbycake committed:
Added the ability to remove task

sebbycake committed:
Refactored Duke class

sebbycake committed:
Refactored switch stmt with enums, and updated test case

sebbycake committed:
Remove enum types to avoid more errors and update test case

sebbycake committed:
Add the ability to save tasks to a text file locally

sebbycake committed:
Change date from String to LocalDate objects, together with the input format

sebbycake committed:
Merge branch 'branch-Level-8'

sebbycake committed:
Extract more individual classes from Duke

sebbycake committed:
Organize classes into packages

sebbycake committed:
Add unit test cases for Task-related classes

sebbycake committed:
Generate jar files

sebbycake committed:
Add Command classes and fix folder not found error

sebbycake committed:
Add javadocs

sebbycake committed:
Reformat code to meet coding standards

sebbycake committed:
Add search feature using a kwyword as query

sebbycake committed:
Resolve merge conflict in javadocs

sebbycake committed:
Resolve merge conflict by adding new code from level 9 branch

sebbycake committed:
Merge branch 'add-gradle-support' of https://github.com/sebbycake/ip

sebbycake committed:
Add checkstyle and fix coding style errors

sebbycake committed:
Fix write tasks  to txt file error

aikenwx reviewed (commented):
LGTM. Very neat and readable code.

sebbycake committed:
Add GUI for Duke

sebbycake committed:
Update build.gradle to generate JAR file and add greeting message to GUI

sebbycake committed:
Add and handle more exceptions

jaredlhf commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,41 @@
+
+    private Duke duke;
+
+    {
+        try {
+            duke = new Duke("data/duke.txt");
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
```
Whats the reason for the extra set of curly brackets?

jaredlhf commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,118 @@
+     */
+    public static Command parse(String userInput) throws DukeException {
+        String command = getCommand(userInput);
+        switch (command) {
+        case "todo":
+            return new AddCommand(
+                    getDescription(userInput, "todo", "ignore"),
+                    "todo");
+        case "deadline":
+            return new AddCommand(
+                    getDescription(userInput, "deadline ", "/by "),
+                    getDate(userInput, "/by "),
+                    "deadline"
+            );
+        case "event":
+            return new AddCommand(
+                    getDescription(userInput, "event ", "/at "),
+                    getDate(userInput, "/at "),
+                    getTime(userInput),
+                    "event"
+            );
+        case "find":
+            return new FindCommand(getSearchQuery(userInput));
+        case "done":
+            return new DoneCommand(getTaskNumber(userInput));
+        case "delete":
+            return new DeleteCommand(getTaskNumber(userInput));
+        case "list":
+            return new ListCommand();
+        case "bye":
+            return new ExitCommand();
+        default:
+            return new InvalidCommand();
```
I like that you used a switch statement here!

jaredlhf commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,146 @@
+     */
+    public List<Task> filter(String keyword) {
+        String regex = ".*\\b" + keyword + "\\b.*";
+        return this.tasksList
+                .stream()
+                .filter(task -> task.getDescription().matches(regex))
+                .collect(Collectors.toList());
```
I like the appropriate line breaks used here!

jaredlhf commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,40 @@
+public class Duke {
+
+    private Storage storage;
+    private TaskList tasks;
```
tasks might be a misleading name for a TaskList object, perhaps a more specific name would be better?

sebbycake replied:
Sure, thanks!

jaredlhf reviewed (commented):
Overall, LGTM. Code is neat and easy to read. Good use of variable names and follows the coding standards.

sebbycake committed:
JavaFX styling: use CSS file to manage styles in one place

It is difficult to manage new styles using inline style within each Node element in fxml. As more styles are added, the code becomes long and unreadable.

In this commit, I removed border radius for text input and button components such that they fit the window size nicely. I also cropped the user and duke image to be circular which fits most modern messenger platforms.

sebbycake committed:
Add Java assert statements to test assumptions

sebbycake committed:
Modify code to adhere to code quality guidelines

sebbycake committed:
Merge pull request #1 from sebbycake/branch-A-Assertions

Add Java assert statements to test assumptions

sebbycake committed:
Merge pull request #2 from sebbycake/branch-A-CodeQuality

Modify code to adhere to code quality guidelines

sebbycake committed:
Set up CI using GitHub Actions

sebbycake committed:
Make task search more flexible by matching keyword partially

sebbycake committed:
Change profile pictures, add different background colors for dialog

sebbycake committed:
Update README and add Ui image

sebbycake committed:
Add more unit tests

sebbycake committed:
Update Label size and reduce photo size

sebbycake committed:
Update Ui image
